### ISSUE TITLE
Error when trying to use a Global Temporary Table

### ISSUE THREAD

**Issue by alan-rosenthal** [utPLSQL/utPLSQL#462]
Hi - If the procedure that I am testing uses a Global Temporary Table then I get an error when I try to test it.  I get the same error if I try to use a GTT in the test suite to store expected results (which I would then compare to the actual table).

**ORA-19202: Error occurred in XML processing
ORA-08103: object no longer exists
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 105
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 30
ORA-06512: at "UT3.UT_EQUAL", line 70
ORA-06512: at "UT3.UT_EXPECTATION_REFCURSOR", line 21**

When I change these to regular tables the test suite runs fine (without changing any code).

Any idea about this error/bug?

Thanks,
Alan Rosenthal
alan.h.rosenthal@gmail.com

P.S.  I LOVE utPLSQL !!!!!!!!!!!!!

**Comment by jgebal** [comment]
Hi @alan-rosenthal 
Can you give the following info?

- DB version
- utPLSQL version
- Schema owning the test
- Schema owning the tested code
- Schema running the tests
- Isolated example of plsql code that fails with this exception

Thanks
Jacek

**Comment by alan-rosenthal** [comment]
Hi Jacek,

Thanks for the very quick response.  Unfortunately, we use a lot of GTT's in our applications so this is a show stopper.  Let me know what you think?  Maybe I am doing something wrong???

And apologies for the poor code formatting, below.  I had a little trouble with this text editor.


DB Version - 11.2.0.4 (linux) and 12.2.0.1 (windows)
utPLSQL version - 3.0.3
Run from - Toad, SQL Developer, and sqlplus command line
Schema running the tests - arosenthal
Schema with utPLSQL code - ut3

This minimal procedure inserts a row into a table and trims and converts the string to uppercase.
    CREATE OR REPLACE PROCEDURE insert_row(p_n NUMBER, p_d DATE, p_v VARCHAR2)
    IS
    BEGIN
           INSERT INTO four(n, d,  v) VALUES (p_n, p_d, TRIM(UPPER(p_v)));
    END insert_row;
    /

And this test suite spec:
    CREATE OR REPLACE PACKAGE test_insert_row_proc
    IS
       -- %suite (*****     Test INSERT_ROW procedure)
       -- %test(**********     Test empty table)
       PROCEDURE empty_table;

       -- %test(**********     Test 1 row in table, column by column approach)
       PROCEDURE one_row;
   
       -- %test(**********     Test 3rows in table, cursor data embedded in package approach)
       PROCEDURE three_rows;
   
       -- %test(**********     Test 3rows in table, comparing to temp test table approach)
       PROCEDURE five_rows;
END test_insert_row_proc;
/

And this test suite body:
    CREATE OR REPLACE PACKAGE body test_insert_row_proc
    IS
    
	    ----------------------------------------------------------------------------
    	--	Procedure empty_table 
	    ----------------------------------------------------------------------------
    	PROCEDURE empty_table 
	    IS
          n     NUMBER;
          
	    BEGIN
		    SELECT COUNT(*)
		      INTO n
    		  FROM four;
		    ut.expect( n ).to_equal( 0 );
    		
	    END empty_table;
    	
	    -----------------------------------------------------------------------------
    	-- Procedure one_row
	    -----------------------------------------------------------------------------
    	PROCEDURE one_row
	    IS
          n     NUMBER;
          v_n   four.n%TYPE;
          v_d   four.d%TYPE;
          v_v   four.v%TYPE;
          
	    BEGIN
          insert_row(3.14159, TO_DATE('15-JUN-1959', 'DD-MON-YYYY'), 'Happy Birthday');
          
          SELECT COUNT(*)
            INTO n
            FROM four;
          ut.expect( n ).to_equal( 1 );
          
          SELECT n, d, v
            INTO v_n, v_d, v_v
            FROM four;
          ut.expect( v_n ).to_equal( 3.14159 );
          ut.expect( v_d ).to_equal( TO_DATE('15-JUN-1959', 'DD-MON-YYYY') );
          ut.expect( v_v ).to_equal( 'HAPPY BIRTHDAY' ); 
          
       END one_row;
          
	    -----------------------------------------------------------------------------
    	-- Procedure three_rows
	    --    Insert three rows into table
    	-----------------------------------------------------------------------------
	    PROCEDURE three_rows
    	IS
          n           NUMBER;
          actual_cur  sys_refcursor;
          expect_cur  sys_refcursor;
     
	    BEGIN
          insert_row(123, TO_DATE('15-JUN-1959', 'DD-MON-YYYY'), 'Alan');
          insert_row(456, TO_DATE('01-MAY-1961', 'DD-MON-YYYY'), 'Ellin');
          insert_row(789, TO_DATE('22-NOV-1990', 'DD-MON-YYYY'), 'Sarah');
    
          OPEN actual_cur FOR SELECT n, d, v 
                                FROM four 
                            ORDER BY d;
                            
          OPEN expect_cur FOR SELECT n, d, v
                                FROM (SELECT 123 as n, TO_DATE('15-JUN-1959', 'DD-MON-YYYY') as d, 'ALAN' as v FROM DUAL
                                      UNION
                                      SELECT 456 as n, TO_DATE('01-MAY-1961', 'DD-MON-YYYY') as d, 'ELLIN' as v FROM DUAL
                                      UNION
                                      SELECT 789 as n, TO_DATE('22-NOV-1990', 'DD-MON-YYYY') as d, 'SARAH' as v FROM DUAL)
                             ORDER BY d;
          ut.expect( actual_cur ).to_equal( expect_cur );     
                
       END three_rows;
    
	    -----------------------------------------------------------------------------
    	-- Procedure five_rows
	    --    Insert five_rows into table
    	-----------------------------------------------------------------------------
	    PROCEDURE five_rows
    	IS
          n           NUMBER;
          actual_cur1  sys_refcursor;
          expect_cur1  sys_refcursor;
     
	    BEGIN
          INSERT INTO four_test ( n, d, v ) VALUES (123, TO_DATE('15-JUN-1959', 'DD-MON-YYYY'), 'ALAN');
          INSERT INTO four_test ( n, d, v ) VALUES (456, TO_DATE('01-MAY-1961', 'DD-MON-YYYY'), 'ELLIN');
          INSERT INTO four_test ( n, d, v ) VALUES (789, TO_DATE('19-Jun-1995', 'DD-MON-YYYY'), 'RACHEL');
          INSERT INTO four_test ( n, d, v ) VALUES (012, TO_DATE('22-NOV-1990', 'DD-MON-YYYY'), 'SARAH');
          INSERT INTO four_test ( n, d, v ) VALUES (345, TO_DATE('04-Sep-1992', 'DD-MON-YYYY'), 'EMILY');
          
          insert_row(123, TO_DATE('15-JUN-1959', 'DD-MON-YYYY'), ' Alan      ');
          insert_row(456, TO_DATE('01-MAY-1961', 'DD-MON-YYYY'), '  Ellin    ');
          insert_row(789, TO_DATE('19-Jun-1995', 'DD-MON-YYYY'), '   Rachel  ');
          insert_row(012, TO_DATE('22-NOV-1990', 'DD-MON-YYYY'), '    Sarah  ');
          insert_row(345, TO_DATE('04-Sep-1992', 'DD-MON-YYYY'), '     Emily ');
          
          OPEN actual_cur1 FOR SELECT n, d, v 
                                FROM four 
                            ORDER BY d;
                            
          OPEN expect_cur1 FOR SELECT n, d, v
                                FROM four_test
                            ORDER BY d;
                             
          ut.expect( actual_cur1 ).to_equal( expect_cur1 );     
                
       END five_rows;	
    
    
    END test_insert_row_proc;
    /

Now, if I use this regular table, the tests run fine:
Using this regular table:
    CREATE TABLE four
    (
       n NUMBER,
       d DATE,
       v VARCHAR2(100 BYTE)
    );

But if I change the table to a GTT, the tests fail:
*****     Test INSERT_ROW procedure
**********     Test empty table [.025 sec]
**********     Test 1 row in table, column by column approach [.003 sec]
**********     Test 3rows in table, cursor data embedded in package approach [.002 sec] (FAILED - 1)
**********     Test 3rows in table, comparing to temp test table approach [.002 sec] (FAILED - 2)
Failures:
1) three_rows
ORA-19202: Error occurred in XML processing
ORA-08103: object no longer exists
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 95
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 21
ORA-06512: at "UT3.UT", line 79
ORA-06512: at "AROSENTHAL.TEST_INSERT_ROW_PROC", line 72
ORA-06512: at line 6
2) five_rows
ORA-19202: Error occurred in XML processing
ORA-08103: object no longer exists
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 95
ORA-06512: at "UT3.UT_DATA_VALUE_REFCURSOR", line 21
ORA-06512: at "UT3.UT", line 79
ORA-06512: at "AROSENTHAL.TEST_INSERT_ROW_PROC", line 107
ORA-06512: at line 6
Finished in .033971 seconds
4 tests, 0 failed, 2 errored, 0 disabled, 0 warning(s)

**Comment by alan-rosenthal** [comment]
Hi Again,

I forgot to include the DDL for the four_test table.  I was trying to use this to put my expected results in and then compare a query of that expected table to the actual query from the application table.

I get the same results if I make this a regular table (it works) or a GTT (it doesn't work).

    CREATE TABLE FOUR_TEST
    (
      N  NUMBER,
      D  DATE,
      V  VARCHAR2(100 BYTE)
    );

**Comment by jgebal** [comment]
Hi Alan,

1. You cannot be using v3.0.3 as it was not yet relased ;). I guess you're on v3.0.2 or you are using a for of our `develop` branch. Can you give me the results of: `select ut.version() from dual;`
2. Why are you prefixing your test names with ****? You can simply use:
```sql
set serveroutput on size unlimited format truncated
exec ut.run(....);
```
If you are running from tools like TOAD, there is an option to disable trimming leading spaces. SQL Developer doesn't require anything like this. 

Not using `****` will make your life much easier.

Also, you might want to have a look at `utplsql-cli` for running tests from command like (bash/windows) if you are using any CI/CD servers like Jenkins to run your tests periodically.

**Comment by jgebal** [comment]
This is actually a bug/limitation of current implementation.
It's related to the fact that we use autonomous transaction in cursor processing. It seems wrong and we should remove the autonomous transaction from the cursor handling routine.

**Comment by jgebal** [comment]
@alan-rosenthal 
Thank you for reporting this issue. We will work on fixing it.

**Comment by pesse** [comment]
@jgebal Is there a particular reason autonomous transaction was used?
If not I'll create an auto-test and fix for this problem, for cursor comparison seems to be my topic ;)

**Comment by jgebal** [comment]
Hi @pesse
Please go ahead and remove autonomous transactions.
Just use the new approach for self-testing. If you need any additional info, let me know.
Thanks
Jacek

**Comment by jgebal** [comment]
Hi Alan, please have a look at this build job.
https://travis-ci.org/utPLSQL/utPLSQL/jobs/266775945#L5655
It runs documentation reporter in colour and reports code coverage to sonar.
You don't need * in your test names as reporter is doing indentation for you. You just need proper format of DBMS_OUTPUT ( without trimming leading spaces).
If you prefer * over space, I'd suggest writing custom reporter on top on documentation reporter.

**Comment by jgebal** [comment]
This fix will be partof v3.0.3 release

**Comment by jgebal** [comment]
Hi @alan-rosenthal v3.0.3 is released. Your issue is now fixed.
